### PR TITLE
chore: add support for Kubernetes v1.21.10 and v1.22.7 on Azure Stack Hub

### PR DESCRIPTION
**Pull request by haofan-ms**
**Reason for Change**:

Update external-cloud-provider and azuredisk-csi-driver configurations to enable support for K8s 1.21.10 and 1.22.7 on Azure Stack Hub

**Issue Fixed**:

**Credit Where Due**:

Does this change contain code from or inspired by another project?

- [ ] No
- [ ] Yes

If "Yes," did you notify that project's maintainers and provide attribution?

- [ ] No
- [ ] Yes

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

**Comment by jadarsie** [comment]
Please include brief description of what you are pushing, why we care and how it is accomplished.

File: pkg/api/components.go [hunk 1]
```diff
@@ -270,7 +270,7 @@ func getComponentDefaultContainerImage(component string, cs *ContainerService) s
 	k8sComponents := GetK8sComponentsByVersionMap(kubernetesConfig)[cs.Properties.OrchestratorProfile.OrchestratorVersion]
 	hyperkubeImageBase := kubernetesImageBase
 	hyperkubeImage := hyperkubeImageBase + k8sComponents[common.Hyperkube]
-	if cs.Properties.IsAzureStackCloud() {
+	if cs.Properties.IsAzureStackCloud() && !common.IsKubernetesVersionGe(cs.Properties.OrchestratorProfile.OrchestratorVersion, "1.21.0") {
```
Comment (jadarsie) [inline pkg/api/components.go:273]: include please a small comment explaining why the behavior changes starting with 1.21

File: pkg/api/types.go [hunk 2]
```diff
@@ -1232,6 +1232,9 @@ func (p *Properties) ShouldEnableAzureCloudAddon(addonName string) bool {
 	if !to.Bool(o.KubernetesConfig.UseCloudControllerManager) {
 		return false
 	}
+	if addonName == common.AzureDiskCSIDriverAddonName && p.IsAzureStackCloud() {
```
Comment (jadarsie) [inline pkg/api/types.go:1235]: Update the method documentation please:

`// For Azure Stack Hub clusters, ...`

File: pkg/api/components.go (hunk 1 above)
Comment (jadarsie) [reply pkg/api/components.go:273]: the alternative is to write code that document itself

```
variableWithMeaningfulName := !common.IsKubernetesVersionGe(...)
if cs.Properties.IsAzureStackCloud() && variableWithMeaningfulName { ... }
```

File: test/e2e/cluster.sh [hunk 3]
```diff
@@ -125,6 +126,108 @@ function rotateCertificates {
...
+    -e CUSTOM_KUBE_PROXY_IMAGE=${CUSTOM_KUBE_PROXY_IMAGE} \
+    -e CUSTOM_KUBE_APISERVER_IMAGE=${CUSTOM_KUBE_APISERVER_IMAGE} \
+    -e CUSTOM_KUBE_SCHEDULER_IMAGE=${CUSTOM_KUBE_SCHEDULER_IMAGE} \
+    -e CUSTOM_KUBE_CONTROLLER_MANAGER_IMAGE=${CUSTOM_KUBE_CONTROLLER_MANAGER_IMAGE} \
+    -e CUSTOM_KUBE_BINARY_URL=${CUSTOM_KUBE_BINARY_URL} \
+    -e CUSTOM_WINDOWS_PACKAGE_URL=${CUSTOM_WINDOWS_PACKAGE_URL} \
+    -e AZURE_CORE_ONLY_SHOW_ERRORS="True" \
+    "${DEV_IMAGE}" make test-kubernetes || exit 1
+
+  echo "Installing azuredisk-csi-driver..."
```
Comment (jadarsie) [inline test/e2e/cluster.sh:221]: I think this is unnecessary as you can leverage `cluster-init` in this case: https://github.com/Azure/aks-engine/blob/master/docs/topics/clusterdefinitions.md#components

File: test/e2e/cluster.sh [hunk 4]
```diff
@@ -169,95 +272,155 @@ else
...
--e KAMINO_VMSS_PROTOTYPE_IMAGE_TAG=${KAMINO_VMSS_PROTOTYPE_IMAGE_TAG} \
--e CUSTOM_KUBE_PROXY_IMAGE=${CUSTOM_KUBE_PROXY_IMAGE} \
--e CUSTOM_KUBE_APISERVER_IMAGE=${CUSTOM_KUBE_APISERVER_IMAGE} \
--e CUSTOM_KUBE_SCHEDULER_IMAGE=${CUSTOM_KUBE_SCHEDULER_IMAGE} \
--e CUSTOM_KUBE_CONTROLLER_MANAGER_IMAGE=${CUSTOM_KUBE_CONTROLLER_MANAGER_IMAGE} \
--e CUSTOM_KUBE_BINARY_URL=${CUSTOM_KUBE_BINARY_URL} \
--e CUSTOM_WINDOWS_PACKAGE_URL=${CUSTOM_WINDOWS_PACKAGE_URL} \
--e AZURE_CORE_ONLY_SHOW_ERRORS="True" \
-"${DEV_IMAGE}" make test-kubernetes || tryExit && renameResultsFile "deploy"
+
```
Comment (jadarsie) [inline test/e2e/cluster.sh:275]: I think this can be undone if you use `cluster-init`

File: test/e2e/cluster.sh [hunk 5]
```diff
@@ -584,6 +747,15 @@ if [ "${SCALE_CLUSTER}" = "true" ]; then
 fi
 
 if [ "${UPGRADE_CLUSTER}" = "true" ]; then
+  if [ "${DEPLOY_AND_INSTALL_AZUREDISKCSI}" = "true" ]; then
```
Comment (jadarsie) [inline test/e2e/cluster.sh:750]: why do we need to uninstall before upgrade?

File: pkg/api/components.go (hunk 1 above)
Comment (haofan-ms) [reply pkg/api/components.go:274]: added comment

File: pkg/api/types.go (hunk 2 above)
Comment (haofan-ms) [reply pkg/api/types.go:1236]: added comment

**Comment by marosset** [comment]
Make sure https://github.com/Azure/aks-engine/blob/93a472345081f4fd401e1a7010e8e5b0baa7485f/vhd/packer/configure-windows-vhd.ps1#L120-L129 gets updated too!

**Review by jadarsie — APPROVED** [review]

**Review by marosset — APPROVED** [review]
powershell changes LGTM

**Pull request merged**